extension/data.ts

```diff
@@ -296,93 +287,308 @@ class RcxDict {
+    ｳ: 'う',
+    ｪ: 'ぇ',
+    ｴ: 'え',
+    ｫ: 'ぉ',
+    ｵ: 'お',
+    ｶ: 'か',
+    ｶﾞ: 'が',
+    ｷ: 'き',
+    ｷﾞ: 'ぎ',
+    ｸ: 'く',
+    ｸﾞ: 'ぐ',
+    ｹ: 'け',
+    ｹﾞ: 'げ',
+    ｺ: 'こ',
+    ｺﾞ: 'ご',
+    ｻ: 'さ',
+    ｻﾞ: 'ざ',
+    ｼ: 'し',
+    ｼﾞ: 'じ',
+    ｽ: 'す',
+    ｽﾞ: 'ず',
+    ｾ: 'せ',
+    ｾﾞ: 'ぜ',
+    ｿ: 'そ',
+    ｿﾞ: 'ぞ',
+    ﾀ: 'た',
+    ﾀﾞ: 'だ',
+    ﾁ: 'ち',
+    ﾁﾞ: 'ぢ',
+    ｯ: 'っ',
+    ﾂ: 'つ',
+    ﾂﾞ: 'づ',
+    ﾃ: 'て',
+    ﾃﾞ: 'で',
+    ﾄ: 'と',
+    ﾄﾞ: 'ど',
+    ﾅ: 'な',
+    ﾆ: 'に',
+    ﾇ: 'ぬ',
+    ﾈ: 'ね',
+    ﾉ: 'の',
+    ﾊ: 'は',
+    ﾊﾞ: 'ば',
+    ﾊﾟ: 'ぱ',
+    ﾋ: 'ひ',
+    ﾋﾞ: 'び',
+    ﾋﾟ: 'ぴ',
+    ﾌ: 'ふ',
+    ﾌﾞ: 'ぶ',
+    ﾌﾟ: 'ぷ',
+    ﾍ: 'へ',
+    ﾍﾞ: 'べ',
+    ﾍﾟ: 'ぺ',
+    ﾎ: 'ほ',
+    ﾎﾞ: 'ぼ',
+    ﾎﾟ: 'ぽ',
+    ﾏ: 'ま',
+    ﾐ: 'み',
+    ﾑ: 'む',
+    ﾒ: 'め',
+    ﾓ: 'も',
+    ｬ: 'ゃ',
+    ﾔ: 'や',
+    ｭ: 'ゅ',
+    ﾕ: 'ゆ',
+    ｮ: 'ょ',
+    ﾖ: 'よ',
+    ﾗ: 'ら',
+    ﾘ: 'り',
+    ﾙ: 'る',
+    ﾚ: 'れ',
+    ﾛ: 'ろ',
+    ﾜ: 'わ',
+    ｦ: 'を',
+    ﾝ: 'ん',
+    ｳﾞ: 'ゔ',
+    '‌': '‌', // the non-joiner used in google docs
+    voicedSoundMark: 'ﾟ',
+    semiVoicedSoundMark: 'ﾞ',
+  };
+
+  isKana(charCode: number): boolean {
+    return (
+      (charCode >= 12352 && charCode <= 12447) ||
+      (charCode >= 12448 && charCode <= 12543) ||
+      (charCode >= 65382 && charCode <= 65437)
+    );
+  }
 
-  cv: number[] = [
-    0x30f4, 0xff74, 0xff75, 0x304c, 0x304e, 0x3050, 0x3052, 0x3054, 0x3056,
-    0x3058, 0x305a, 0x305c, 0x305e, 0x3060, 0x3062, 0x3065, 0x3067, 0x3069,
-    0xff85, 0xff86, 0xff87, 0xff88, 0xff89, 0x3070, 0x3073, 0x3076, 0x3079,
-    0x307c,
-  ];
-  cs: number[] = [0x3071, 0x3074, 0x3077, 0x307a, 0x307d];
+  convertToHiragana(str: string): string {
+    let result = '';
+    for (let i = 0; i < str.length; i++) {
+      const char = str.charAt(i);
+      const nextChar = i + 1 <= str.length - 1 ? str.charAt(i + 1) : null;
+      const semiVoiced =
+        nextChar === this.hiraganaLookup['semiVoicedSoundMark'];
+      const voiced = nextChar === this.hiraganaLookup['voicedSoundMark'];
+      const key = semiVoiced
+        ? char + nextChar
+        : voiced
+        ? char + nextChar
+        : char;
+      const hiragana = this.hiraganaLookup[key];
+      result += hiragana !== undefined ? hiragana : char;
+
+      if (semiVoiced || voiced) {
+        i++;
+      }
+    }
+    return result;
+  }
 
   wordSearch(
     word: string,
     doNames: boolean,
     max?: number
   ): DictEntryData | null {
     let i;
-    let u;
-    let v;
-    let reason: string;
-    let p;
     const trueLen = [0];
     const entry = RcxDict.createDefaultDictEntry();
-
-    // half & full-width katakana to hiragana conversion
-    // note: katakana vu is never converted to hiragana
-
-    p = 0;
-    reason = '';
-    for (i = 0; i < word.length; ++i) {
-      u = v = word.charCodeAt(i);
-
-      // Skip Zero-width non-joiner used in Google Docs between every
-      // character.
-      if (u === 8204) {
```

Comment:
I couldn't tell but do you still have the logic which skips zero-width non-joiner? I saw it was in your table but didn't see it referenced your new code.

Reply:
Yeah, I have it in the lookup table which maps to the same character (essentially skipping / not converting it)

Reply:
But iiuc, the important part of of this check is not that we don't convert it but that it's removed from the final word after conversion. The way this code works is that at the end of each loop the new converted character is appended to the `reason` variable and then when all characters are converted `word` is reset to the new `reason` value.

if we're just converting it directly then the final word we use to do the search will still have this useless character in it, right?

I think part of the problem is that it's shoehorned in with the conversion logic; it might be better to have a separate part of the code which strips out unneeded characters and then group them up in something like a `normalize` method...

Reply:
You are 100% right. This could cause issues with the lengths. I just did a sample test and event though it's invisible, it gets counted with `.length`